module3/lessons/api_authorization.md

```diff
@@ -0,0 +1,502 @@
+
+We are using ruby's [SecureRandom](https://ruby-doc.org/stdlib-3.1.1/libdoc/securerandom/rdoc/SecureRandom.html) module to generate individual authentication tokens that are tied to specific users.  This is not the *only* way to generate authentication tokens - can you imagine a few other ways that we might generate tokens?
+
+<section class='dropdown'>
+### 3 common strategies
+
+**Randomly Generated Token (UUID or Secure Random)**
+How it works: A random string is generated using libraries like SecureRandom in Ruby or generating a UUID (Universally Unique Identifier). These tokens are unique, hard to guess, and often stored in the database associated with the user.
+
+Example in Ruby:
+```ruby
+token = SecureRandom.hex(20)
+```
+
+Advantages:
+* Simple to implement.
+* Tokens are unique and can have a long enough length to ensure security.
+Drawbacks:
+* Tokens don’t expire unless explicitly set to do so.
+* If a token is compromised, it remains valid until revoked manually or through the system.
+
+**JWT (JSON Web Tokens)**
+How it works: JWTs are a compact, URL-safe token that consist of three parts: a header, payload, and signature. The payload contains the user data (claims) and the signature ensures the integrity of the token. JWTs can be signed with a secret key (HS256) or a public/private key pair (RS256).
+Example in Ruby (using jwt gem):
+
+```ruby
+payload = { user_id: user.id, exp: (Time.now + 1.day).to_i }
+token = JWT.encode(payload, 'secret_key', 'HS256')
+```
+
+Advantages:
+* Stateless authentication (no need to store tokens in the database).
+* Built-in support for expiration (exp claim) and can carry user data.
+* Easily decodable for quick verification of claims.
+Drawbacks:
+* If not handled securely, the payload can be tampered with (though the signature protects this).
+* Cannot be revoked easily since tokens are not stored in the database.
+
+**OAuth Access Tokens**
+How it works: OAuth is an open-standard authorization framework that issues access tokens. The client receives a token after authenticating, which is then used to access API endpoints. Access tokens are usually short-lived, with a refresh token issued to extend access.
+
+Example: In Rails, you could implement this using libraries like Doorkeeper to handle OAuth token generation and expiration.
+
+Advantages:
+* Token expiration and refresh mechanisms are built-in.
+* Secure for third-party access delegation.
+* Standardized and well-supported in libraries.
+Drawbacks:
+* More complex to set up than random tokens or JWTs.
+* Tokens can expire frequently, requiring refresh logic.
+* Each strategy has different use cases, with random tokens being simple but less flexible, JWTs offering stateless and efficient token handling, and OAuth providing a comprehensive authorization framework, especially for third-party integrations.
+
```

Comment:
This is a great level of abstraction of the different types of tokens IMO! Good exposure without getting into the weeds :)